runtime/src/main/kotlin/kotlin/native/concurrent/MutableData.kt

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.native.concurrent
+
+import kotlin.native.internal.*
+import kotlinx.cinterop.*
+
+@SymbolName("Kotlin_Any_share")
+external private fun Any.share()
+
+@SymbolName("Kotlin_CPointer_CopyMemory")
+external private fun CopyMemory(to: COpaquePointer?, from: COpaquePointer?, count: Int)
+
+/**
+ * Mutable concurrently accessible data buffer. Could be accessed from several workers simulteniously.
+ */
+@Frozen
+public class MutableData {
+    init {
+        // Instance of MutableData is shared.
+        share()
+    }
+
+    private var buffer = ByteArray(16).apply { share() }
```

Comment:
Please note that buffer of size `0` is potentially dangerous since `buffer.usePinned { it.addressOf(0) }` will crash AFAIK. So zero capacity need to be forbidden or need to be checked everywhere